File: paradigms/clojure-solutions/functionalExpression.clj
; HW 9: Functional expressions
(defn operation [f] (fn [& operands] (fn [vars] (apply f (map #(% vars) operands)))))

(defn constant [number] (fn [vars] number))
(defn variable [name] (fn [vars] (vars name)))

(def multiply (operation *))
(def subtract (operation -))
(def divide (operation (fn [a b] (/ (double a) (double b)))))
(def add (operation +))
(def negate (operation -))
(def sin (operation (fn [a] (Math/sin a))))
(def cos (operation (fn [a] (Math/cos a))))

(def signToOpFunctional {
              '+ add,
              '- subtract,
              '/ divide,
              'negate negate,
              '* multiply,
              'sin sin,
              'cos cos
               })

(defn fromListToFunctionalExpr [expressionList] (cond
                                        (list? expressionList) (
                                            apply (signToOpFunctional (first expressionList)) (map fromListToFunctionalExpr (rest expressionList))
                                            )
                                        (number? expressionList) (constant expressionList)
                                        (symbol? expressionList) (variable (str expressionList))
                                        ))
(defn parseFunction [expression] (fromListToFunctionalExpr (read-string expression)))

File: paradigms/javascript-solutions/functionalExpression.js
const operation = f => (...operands) => (...values) => operands.length > 1
    ? f(operands[0](...values), operands[1](...values))
    : f(operands[0](...values));

const subtract = operation((a, b) => a - b);
const add = operation((a, b) => a + b);
const multiply = operation((a, b) => a * b);
const divide = operation((a, b) => a / b);
const negate = operation(a => -a);
const cnst = value => () => value;

const variable = name => (...values) => {
    switch (name) {
        case "x":
            return values[0];
        case "y":
            return values[1];
        case "z":
            return values[2];
    }
}

